File: 001-099/02/2.js
var a = 0, b = 1;
var total = 0, tmp;
while (b < 4000000) {
    tmp = a, a = b;
    b += tmp;
    if (b % 2 == 0) {
        total += b;
    }
}
console.log(total);

File: 001-099/02/2.rs
struct Fib {
    a: u64,
    b: u64,
    limit: u64,
}

impl Iterator for Fib {
    type Item = u64;
    fn next(&mut self) -> Option<u64> {
        let tmp = self.a;
        self.a = self.b;
        self.b += tmp;
        if self.a < self.limit {
            Some(self.a)
        } else {
            None
        }
    }
}

fn main() {
    let f = Fib {a: 1, b: 1, limit: 4000000};
    let s = f.filter(|x| x % 2 == 0).fold(0, |a, b| a + b);
    println!("{}", s);
}
